cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/loader/LogAppenderLoaderService.java

```diff
@@ -0,0 +1,137 @@
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import com.google.common.collect.Iterators;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.inject.Inject;
+import org.elasticsearch.common.Strings;
+import org.slf4j.ILoggerFactory;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * Service that programmatically loads additional log appenders. It provides classloader isolation for additional
+ * appender.
+ */
+public class LogAppenderLoaderService extends AbstractIdleService {
+  private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(LogAppenderLoaderService.class);
+  private final CConfiguration cConf;
+  private Appender<ILoggingEvent> logAppender;
+
+  @Inject
+  public LogAppenderLoaderService(CConfiguration cConf) {
+    this.cConf = cConf;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    String provider = cConf.get(Constants.Logging.LOG_APPENDER_PROVIDER);
+    // If there is no provider, there is nothing to load. This can happen if no additional appenders are provided.
+    if (Strings.isNullOrEmpty(provider)) {
+      return;
+    }
+
+    logAppender = loadLogAppender(provider, cConf);
+    initialize(logAppender, cConf);
+
+    LOG.info("Log Appender {} is initialized and started.", provider);
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    if (logAppender != null) {
+      logAppender.stop();
+    }
+  }
+
+  /**
+   * Loads log appender using extension loader.
+   */
+  @SuppressWarnings("unchecked")
+  private Appender<ILoggingEvent> loadLogAppender(String appenderProvider, CConfiguration cConf) {
+    Appender<ILoggingEvent> appender = new LogAppenderExtensionLoader(cConf).get(appenderProvider);
+    if (appender == null) {
+      // this will not happen unless log appender provider is misconfigured.
+      throw new RuntimeException("Log appender " + appenderProvider + " is not constructed. " +
+                                   "Please provide correct log appender provider.");
+    }
+    return appender;
+  }
+
+  /**
+   * Initializes log appender by reading properties from cConf.
+   */
+  private void initialize(Appender<ILoggingEvent> appender, CConfiguration cConf) throws Exception {
+    ILoggerFactory loggerFactory = LoggerFactory.getILoggerFactory();
+    LoggerContext loggerContext = (LoggerContext) loggerFactory;
+    if (loggerContext != null) {
+      Logger logger = loggerContext.getLogger(Logger.ROOT_LOGGER_NAME);
+      // Check if the logger already contains the logAppender
+      if (!Iterators.contains(logger.iteratorForAppenders(), appender)) {
+        appender.setContext(loggerContext);
+        // Display any errors during initialization of log appender to console
+        configureStatusManager(loggerContext);
+
+        ClassLoader oldClassLoader = getClass().getClassLoader();
```

Comment:
You can combine these two lines as `ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(appender.getClass().getClassLoader());`